nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/lambda/PutLambda.java

```diff
@@ -0,0 +1,239 @@
+            }
+
+            if ( ! StringUtils.isBlank(result.getFunctionError()) ){
+                flowFile = session.putAttribute(flowFile, AWS_LAMBDA_RESULT_FUNCTION_ERROR, result.getFunctionError());
+                session.transfer(flowFile, REL_FAILURE);
+            } else {
+                session.transfer(flowFile, REL_SUCCESS);
+                final long totalTimeMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
+                session.getProvenanceReporter().send(flowFile, functionName, totalTimeMillis);
+            }
+        } catch (final InvalidRequestContentException
+            | InvalidParameterValueException
+            | RequestTooLargeException
+            | ResourceNotFoundException
+            | UnsupportedMediaTypeException unrecoverableException) {
+                getLogger().error("Failed to invoke lambda {} with unrecoverable exception {} for flow file {}",
+                    new Object[]{functionName, unrecoverableException, flowFile});
+                flowFile = populateExceptionAttributes(session, flowFile, unrecoverableException);
+                session.transfer(flowFile, REL_FAILURE);
+        } catch (final TooManyRequestsException retryableServiceException) {
+            getLogger().error("Failed to invoke lambda {} with exception {} for flow file {}, therefore penalizing flowfile",
+                new Object[]{functionName, retryableServiceException, flowFile});
+            flowFile = populateExceptionAttributes(session, flowFile, retryableServiceException);
+            flowFile = session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
+            context.yield();
+        } catch (final AmazonServiceException unrecoverableServiceException) {
+            getLogger().error("Failed to invoke lambda {} with exception {} for flow file {} sending to fail",
+                new Object[]{functionName, unrecoverableServiceException, flowFile});
+            flowFile = populateExceptionAttributes(session, flowFile, unrecoverableServiceException);
+            session.transfer(flowFile, REL_FAILURE);
+            context.yield();
+        } catch (final Exception exception) {
+            getLogger().error("Failed to invoke lambda {} with exception {} for flow file {}",
+                new Object[]{functionName, exception, flowFile});
+            session.transfer(flowFile, REL_FAILURE);
+            context.yield();
+        }
+    }
+
+    /**
+     * Populate exception attributes in the flow file
+     * @param session process session
+     * @param flowFile the flow file
+     * @param exception exception thrown during invocation
+     * @return FlowFile the updated flow file
+     */
+    private FlowFile populateExceptionAttributes(final ProcessSession session, FlowFile flowFile,
+            final AmazonServiceException exception) {
+        flowFile = session.putAttribute(flowFile, AWS_LAMBDA_EXCEPTION_MESSAGE, exception.getErrorMessage());
```

Comment:
I would recommend that we create a HashMap here to put these attributes into and then call session.putAttributes() instead of many calls to session.putAttribute(), as it is much more efficient. Session.putAttribute() has to create a new FlowFile for each invocation, and this is more expensive than just creating a HashMap once.